src/primary_names.lua

```diff
@@ -0,0 +1,301 @@
+	}
+end
+
+--- Get a primary name request, safely deep copying the request
+--- @param address string
+--- @return PrimaryNameRequest|nil primaryNameClaim - the request found, or nil if it does not exist
+function primaryNames.getPrimaryNameRequest(address)
+	return utils.deepCopy(primaryNames.getUnsafePrimaryNameRequests()[address])
+end
+
+--- Unsafe access to the primary name requests
+--- @return table<string, PrimaryNameRequest> primaryNameClaims - the primary name requests
+function primaryNames.getUnsafePrimaryNameRequests()
+	return PrimaryNames.requests or {}
+end
+
+function primaryNames.getUnsafePrimaryNames()
+	return PrimaryNames.names or {}
+end
+
+--- Unsafe access to the primary name owners
+--- @return table<string, PrimaryName> primaryNames - the primary names
+function primaryNames.getUnsafePrimaryNameOwners()
+	return PrimaryNames.owners or {}
+end
+
+--- Action taken by the owner of a primary name. This is who pays for the primary name.
+--- @param recipient string -- the address that is requesting the primary name
+--- @param from string -- the process id that is requesting the primary name for the owner
+--- @param timestamp number -- the timestamp of the request
+--- @return PrimaryNameWithOwner primaryNameWithOwner - the primary name with owner data
+function primaryNames.approvePrimaryNameRequest(recipient, name, from, timestamp)
+	local request = primaryNames.getPrimaryNameRequest(recipient)
+	assert(request, "Primary name request not found")
+	assert(request.endTimestamp > timestamp, "Primary name request has expired")
+
+	-- assert the process id in the initial request still owns the name
+	local record = arns.getRecord(request.baseName)
+	assert(record, "ArNS record '" .. request.baseName .. "' does not exist")
+	assert(record.processId == from, "Primary name request must be approved by the owner of the base name")
+
+	-- assert the name matches the request
+	assert(request.name == name, "Provided name does not match the primary name request")
+
+	-- set the primary name
+	local newPrimaryName = primaryNames.setPrimaryNameFromRequest(recipient, request, timestamp)
+	return newPrimaryName
+end
+
+--- Update the primary name maps and return the primary name. Removes the request from the requests map.
+--- @param recipient string -- the address that is requesting the primary name
+--- @param request PrimaryNameRequest
+--- @param startTimestamp number
+--- @return PrimaryNameWithOwner primaryNameWithOwner - the primary name with owner data
+function primaryNames.setPrimaryNameFromRequest(recipient, request, startTimestamp)
+	PrimaryNames.names[request.name] = recipient
+	PrimaryNames.owners[recipient] = {
+		name = request.name,
+		baseName = request.baseName,
+		startTimestamp = startTimestamp,
+	}
+	PrimaryNames.requests[recipient] = nil
+	return {
+		name = request.name,
+		owner = recipient,
+		startTimestamp = startTimestamp,
+		baseName = request.baseName,
+		-- TODO: add base name owner if useful
```

Comment:
I question the utility since the owner of the base name can change at any time without this process knowing.

Reply:
correct, this would be just ensuring we send notices to the correct base name owner. not stored in state.

Reply:
[24a6f10](https://github.com/ar-io/ar-io-network-process/pull/175/commits/24a6f1030bc06c3a1875c446318ff19d613ec6bd)